src/sql/src/pure.rs

```diff
@@ -335,10 +293,19 @@ async fn purify_source_format_single(
             }
         },
         Format::Csv {
-            delimiter,
+            delimiter: _,
             ref mut columns,
         } => {
-            purify_csv(file, connector, *delimiter, columns).await?;
+            if let CsvColumns::Header { names } = columns {
+                match connector {
+                    CreateSourceConnector::S3 { .. } => {
+                        if names.is_empty() {
+                            bail!("CSV WITH HEADER for S3 sources requires specifying the header columns");
```

Comment:
nice! the railroad diagram needs to be adjusted to remove the skip over the header columns: https://materialize.com/docs/sql/create-source/s3/

removed here: https://github.com/MaterializeInc/materialize/blob/main/doc/user/layouts/partials/create-source/format/csv/syntax.html#L1 and fixed here https://github.com/MaterializeInc/materialize/blob/main/doc/user/sql-grammar/sql-grammar.bnf#L182-L193

Reply:
Oh good point, thanks! Fixed.